feat(tests/infra) - interactive mode implementation

sedouard opened the pull request:
Creates a 'nw:test --server' switch which starts tests in interactive mode
using testem's server mode. Qunit-logger connects to the server via
web socket and re-runs tests without restarting the process. Interactive
mode shows nwjs where ci mode hides.

sedouard commented on blueprints/node-webkit/index.js:
```diff
@@ -45,7 +45,7 @@ module.exports = {
       json.main = 'dist/index.html';
       json.window = { width: 960, height: 600 };
-
+      json.devDependencies['socket.io-client'] = '^1.3.5';
```
I wasn't able to directly test this - I think it requires `npm publish` access

brzpegasus replied:
We should be able to remove this dependency here too, given we can use the one that Testem provides.

sedouard commented:
I went with the approach of using the testem server. Basically the idea is that we don't exit the test run from `nwjs` and instead hook into the server's socket.io api that it exposes for browsers in `quint-logger`. When the user changes something the watcher will tell testem to instruct the browser to restart the tests which just causes the browser to reload.

Thoughts?

brzpegasus commented:
Thanks! I will take a look later this morning.

brzpegasus commented:
I've been playing with this locally (still am), but here are some initial thoughts.

First, I appreciate the attempt to have this interact with the testem server. I was originally thinking of just reusing the stuff you had already implemented for `ember nw:test` and just not exit the child process (in the runner) at the end, if we're running in interactive mode. However, making this work with Testem's GUI is even nicer.

There are a few issues to address before this can be merged, and I'm still trying to figure some of those items out myself:

* The tests are running too soon.
  * Currently, the QUnit logger attempts to open a socket connection to the testem server, but by the time the connection is established, some of the tests have already run, and nothing ever gets emitted and reported back to Testem for those tests.
  * You can probably set `QUnit.config.autostart = false` then call `QUnit.start()` once the connection is made, but note we'll have to require users to be on `ember-cli-qunit` 0.3.11 given this [issue](https://github.com/ember-cli/ember-cli-qunit/pull/57).

* I always get a `Failed to load resource: net::ERR_ADDRESS_UNREACHABLE` error for `http://localhost:7357/socket.io/?EIO=3&transport=polling&t=<some-token>` when connecting to the testem server, but the connection does succeed. This might again be related to the server not being ready.

* Rather than introducing a new dependency on `socket.io-client`, couldn't we fetch it from Testem? `http://localhost:7357/socket.io/socket.io.js` should give us what we need, no?

* When the tests are done running, the Testem GUI shows "Waiting for runners..." instead of the results of the tests. But things do show up while the test suite is in progress. Gotta figure out how to make those results stay.

* The stuff that the QUnit logger is `console.log`'ging during the tests is showing up in the testem GUI. We'll probably want to turn that off in this mode. It's only needed for CI.

* I plugged this into my demo app as I already had a set of NW.js-dependent tests written for. I find that if I enable the "Preserve log" option in the console, and then manually refresh the page to have the tests re-run, the log shows that the entire suite actually executes multiple times. The first time everything does pass (woo!), but then something causes the page to reload and eventually the tests fail with some weird error (some node error about not being able to find a file named 'uv_cwd').

I need to spend more time with this, but overall, I dig it! We just need to make it work =).

sedouard commented:
Sorry for the delay here by the way - I'm traveling about but I'm going to put a couple fixes in the next couple days.

brzpegasus commented:
@sedouard No problem! I've been busy as well, but I'm hoping to get back to this and help figure out the remaining items before too long.

sedouard commented:
```
The tests are running too soon.
```
Yup - you're right. I took your idea but I added an environment variable to indicate we're in server mode so that we only start the tests on socket connection for server mode tests. Normally I would think that calling the build command with the 'development' environment mode would pass this in to the `meta` environment data on the built index.html. However it remains `test` as environment. So I did this as a workaround.

```
I always get a Failed to load resource: net::ERR_ADDRESS_UNREACHABLE error for http://localhost:7357/socket.io/?EIO=3&transport=polling&t=<some-token> when connecting to the testem server, but the connection does succeed. This might again be related to the server not being ready.
```
Yeah - but is this that big of a deal? Socket.io's behavior is to retry to connect and it'll try to connect again.

```
Rather than introducing a new dependency on socket.io-client, couldn't we fetch it from Testem? http://localhost:7357/socket.io/socket.io.js should give us what we need, no?
```

Agreed - should we get it from `bower_components` or pull it directly from localhost:7357 at run-time? (Currently this commit I just added does the former)

```
When the tests are done running, the Testem GUI shows "Waiting for runners..." instead of the results of the tests. But things do show up while the test suite is in progress. Gotta figure out how to make those results stay.
```

Here's what I see after my tests run:
[![asciicast](https://asciinema.org/a/6rtbv2koyd1nts9w4zri0597t.png)](https://asciinema.org/a/6rtbv2koyd1nts9w4zri0597t)

```
The stuff that the QUnit logger is console.log'ging during the tests is showing up in the testem GUI. We'll probably want to turn that off in this mode. It's only needed for CI.
```

True - this was intentional I didn't think there was any harm in just outputting the qunit-logger's output to the socket-io connection. I can take this out though.

```
I plugged this into my demo app as I already had a set of NW.js-dependent tests written for. I find that if I enable the "Preserve log" option in the console, and then manually refresh the page to have the tests re-run, the log shows that the entire suite actually executes multiple times. The first time everything does pass (woo!), but then something causes the page to reload and eventually the tests fail with some weird error (some node error about not being able to find a file named 'uv_cwd').
```

Can you try it with the new changes?

brzpegasus commented:
@sedouard Sorry, now I'm delaying! I will check this out soon. The plan is to get this all figured out, merged in, and released by June 16th, because guess who's giving a talk at this year's [Wicked Good Ember](http://wickedgoodember.com/) conference? =) It would be great to share some of this work.

sedouard commented:
Excellent and congrats! Looks like we have a deadline :-). 

I haven't ran the PR o your ember project - hoping it solves the test timing issue. I'm hitting a small issue with number of tests reported by qunit maybe or maybe that doesn't reproduce with you.

brzpegasus commented on test-support/node-webkit/qunit-logger.js:
```diff
@@ -1,8 +1,25 @@
+/* global io */
 // dead stupid script to format test output from nw.js to the console
 import QUnit from 'qunit';
+var socket = io('http://localhost:7357');
```
The host and port should be configurable rather than hard-coded here.

brzpegasus commented:
> When the tests are done running, the Testem GUI shows "Waiting for runners..." instead of the results of the tests. But things do show up while the test suite is in progress.

I checked out the video you linked, and I do get the same behavior, but only when tests pass. Have you been able to check what happens on your end when you have a failing test? That's when the testem GUI reloads to say "Waiting for runners..."

sedouard commented:
so testem has exactly [what we want](https://github.com/airportyh/testem/blob/31e070490cb8835a0b05b4f14fc1a75fe71e130a/public/testem/qunit_adapter.js) which does the interaction with the socket.io api correctly. So I based qunit-logger off of it with the current tap output today for non-interactive mode runs

Can you try this one out?

brzpegasus commented:
It is better in that when tests fail, Testem's console does reflect that, instead of showing "Waiting for runners...".

There's still an issue with the tests running twice. You can see that the page reloads in the NW.js window during the test suite execution, after it finishes (and passes) the first time. The second time that the tests re-run, they fail with some Node error about not being able to find `uv_cwd`. I can look more into it.

In the meantime, there is some refactoring I'd like to see in this PR to clean things up a bit. I'll start adding some inline comments as I go through the code.

brzpegasus commented on lib/commands/nw-test/index.js:
```diff
@@ -34,6 +36,37 @@ module.exports = {
     this.quickTemp.remove(this, '-testsDistNW');
   },
 
+  TestServerBuilder: Builder.extend({
```
I'd love to see this `TestServerBuilder` section removed in favor of using addon hooks to manipulate the build pipeline. From what I understand, the reason why you're extending the builder is to do stuff after the build. I think this can be done more elegantly by modifying the addon's [`index.js`](https://github.com/brzpegasus/ember-cli-node-webkit/blob/v0.2.0/index.js) to have the following hook:

```js
postprocessTree: function(type, tree) {
  var funnel = require('broccoli-funnel');
  var mergeTrees = require('broccoli-merge-trees');
  var replace = require('broccoli-string-replace');

  if (type === 'all' && process.env.EMBER_ENV === 'test') {
    // Update the base URL in `tests/index.html`
    var index = replace(tree, {
      files: ['tests/index.html'],
      pattern: {
        match: /base href="\/"/,
        replacement: 'base href="../"'
      }
    });

    // Copy `tests/package.json` to the output directory
    var testPkg = funnel('tests', {
      files: ['package.json'],
      destDir: '/tests'
    });

    return mergeTrees([tree, index, testPkg], { overwrite: true });
  }

  return tree;
}
```

sedouard replied:
yeah - I think I was just trying to avoid mucking with the application's build process all together. (this will impact `ember build`) I thought keeping the behavior just in `ember nw:test` would have been a good idea (for example changing `tests/index.html` file contents) but then again this was just a guess.

We can do it with the ember-cli add-on hooks. However a slight functional problem that I can't seem to figure out is how to hide or show nwjs. Right now the `tests/package.json` in the blue prints has a `window.show` property in its `package.json` file set to false (for a CI mode experience).

If i use `postprocessTree` the `package.json` file doesn't get copied before the the promise from `Builder.build()` (inside [nw-test index.js](lib/commands/nw-test/index.js) ) is resolved in the and so trying to change this flag (by reaching in the file and changing it via code) doesn't work. The `postprocessTree` hook also doesn't tell us what the command line that invoked the code so I don't think using the broccoli plugin will work either. So there's no way to discern from development and ci test experiences.

I was thinking about changing the source package.json... but I think that would cause a pending change on the file under source control. Open to ideas here...

brzpegasus replied:
> this will impact `ember build`.

I think this is fine, because we're only tweaking the build if the environment being built for is `test`.

> However a slight functional problem that I can't seem to figure out is how to hide or show nwjs.

I'm thinking we can just do away with `window.show`. Hiding it isn't important so long as NW.js doesn't stay running in CI mode. For instance, when you configure Testem to run Chrome or Firefox in CI, those browsers will be visible for the duration of the tests, then close immediately afterwards.

brzpegasus commented on lib/commands/nw-test/index.js:
```diff
@@ -106,9 +213,15 @@ module.exports = {
+    var promise;
+    // server mode will build continuously
+    if (options.server) {
+      options.environment = 'development';
```
Why should this be `development`?

sedouard replied:
My original thought was that the usually when you decide to use the `--server` flag, you're developing code not in a CI environment. However I think I actually added it to see if I could some how capture that environment in the built ember cli's application environment that goes into `index.html`. But I think we should just leave this out and let the user decide what environment to use.

brzpegasus commented on lib/commands/nw-test/index.js:
```diff
@@ -97,8 +129,83 @@ module.exports = {
+        options: options
+      });
+
+      var watchDirOrFiles = fs.readdirSync(_this.project.root);
```
What are you trying to achieve here, by adding all the project directories to the watcher (which, if I read this correctly, would also watch `bower_components`, `node_modules`, `.git`, etc.? The watcher should already be watching all the trees defined by the broccoli build.

sedouard replied:
I thought this, too since it has access to the builder.but if I comment out the `addWatchDir` call no watching occurs. Does that reproduce for you as well?

brzpegasus replied:
The watcher still works great for me. If I change any file in `app/` or `tests/`, the app gets rebuilt and the tests re-run.

sedouard replied:
Hmm are you using watchman or the built-in node watcher? Maybe its my configuration.

brzpegasus replied:
Watchman.

brzpegasus commented on index.js:
```diff
@@ -8,6 +8,7 @@ module.exports = {
 
     app.import('vendor/node-webkit/shim.js', { prepend: true });
     app.import({ development: 'vendor/node-webkit/reload.js' });
+    app.import({ development: '../../node_modules/ember-cli/node_modules/testem/node_modules/socket.io/node_modules/socket.io-client/socket.io.js' });
```
I would remove this import and just bring `socket.io` in via the `contentFor` hook. This will inject the script into `tests/index.html`:

```js
// ./index.js

contentFor: function(type) {
  // TODO Make testem server URL configurable
  var testemServer = 'http://localhost:7357';
  if (type === 'test-body' && process.env.EMBER_ENV === 'test') {
    return '<script src="' + testemServer + '/socket.io/socket.io.js"></script>';
  }
}
```

sedouard replied:
Looks good only issue is that this `TODO` to make the port configurable will be a bit more tricky. My original thought would be to take a `--port` option on the command line and pass that directly to testem within the command handler. 

But at this point, we do not have access to the command line arguments (at least I don't see that anywhere in the [docs](https://github.com/ember-cli/ember-cli/blob/master/ADDON_HOOKS.md#config) so I think we would have to use the application configuration for the type of build?

sedouard replied:
Gonna keep this as a TODO, I think the real problem here is that normally testem will start the browser pointed to the server and port. But since we're serving off of a file we don't really have that option, so the client needs to find its port. I think its a bit less straight forward than I thought...

brzpegasus replied:
Np, we can make it configurable later =).

sedouard replied:
There's actually another hidden issue here. Depending on the socket.io.js served off of the testem server causes issues.

This kind of falls to the same theme that it seems like the testem server mode was really designed to host the website content. It seems like testem assumes the browser will wait around until its ready to load the page (which makes sense, because a traditional browser does). But since we're loading the content locally, loading the socket.io.js file from testem via `<script>` tag may result in a dead link if the timing isn't right.

I don't see how to get around this unless we host index.html on testem, at which point changes how the app is ran all together .

brzpegasus replied:
> I thought that the testem wouldn't start the launcher until the server is already up but seems like the timing isn't guaranteed.

How are you determining that? The launchers shouldn't be initialized until the server is ready according to [this](https://github.com/airportyh/testem/blob/19a7f46bfc631d2cadbdc892a090a1d348a601c3/lib/dev/index.js#L88-L106), unless I'm reading that wrong.

sedouard replied:
Never mind I've been working to long clearly. I had left around the extra `exec` in there so it was actually trying to start nw 2x's I just didn't notice ::doh::

brzpegasus commented on lib/commands/nw-test/index.js:
```diff
@@ -97,8 +129,83 @@ module.exports = {
+      var findNW = require('../../helpers/find-nw');
+      var nwPath = safePath(findNW(_this.project));
+      var testPath = safePath(path.join(options.outputPath, 'tests'));
+      exec(nwPath + ' ' + testPath, { env: { "SERVER_MODE": "TRUE" } });
```
We should wait for the Testem server to be ready before launching NW.js. Testem provides an `on_start` hook to let you execute custom code. I would move the `exec` stuff there:

```js
  testemOptions: function() {
    return {
      // ...
      on_start: function() {
        var findNW = require('../../helpers/find-nw');
        var nwPath = safePath(findNW(_this.project));
        var testPath = safePath(path.join(options.outputPath, 'tests'));
        exec(nwPath + ' ' + testPath, { env: { "SERVER_MODE": "TRUE" } });
      }
    };
  }
```

brzpegasus replied:
Actually, couldn't we invoke `nw` by specifying a launcher? I feel like it's the more correct way to do it with testem:

```js
testemOptions: function() {
  return {
    // ...
    launch_in_dev: ['NW.js'],
    launchers: {
      'NW.js': {
        command: testCommand,
        protocol: 'browser'
      }
    }
  };
}
```

sedouard replied:
I tried using a launcher and I guess I didn't get the option right. It was a bit ago when I wrote it but I think I may have missed a flag or maybe I kept the protocol tap. Anyways works great.

brzpegasus commented:
> There's still an issue with the tests running twice. You can see that the page reloads in the NW.js window during the test suite execution, after it finishes (and passes) the first time. The second time that the tests re-run, they fail with some Node error about not being able to find uv_cwd. I can look more into it.

Ok, so I figured out what was causing the reload. Some of my acceptance tests were creating files into directories that were being watched, so of course, they re-triggered the test suite. /facepalm

brzpegasus commented:
@sedouard I played with this a bunch, and I find that after making the adjustments that I suggested, the `QUnit.config.autostart = false` and `QUnit.start()` are no longer necessary. In fact, leaving them is causing the tests to no longer run atomically for reasons I'm a bit unclear on (i.e. the `testStart` hook for test 2 would execute before the `testDone` hook for test 1).

There are some tweaks to make to the QUnit logger to make it work with the non-interactive test mode and custom TAP reporter again (this PR breaks the current behavior as details of failed tests are no longer displayed in CI mode). However, I'll be happy to make those tweaks after I merge this in.

So, I would say the only things holding back this PR are the comments I made. Once those are addressed, we should be good to go (and I did verify locally that with the changes I suggested, everything does work perfectly). Getting close!

rwjblue commented:
@brzpegasus - I am not 100% sure if it is related, but ember-cli-qunit does "things" with autostart (to help mitigate the sourcemap issue).  See [here](https://github.com/ember-cli/ember-cli-qunit/blob/c775d09ee4c5122e6b9c9e7fb22a734ac3628e67/vendor/ember-cli-qunit/test-loader.js#L16-L25).

brzpegasus commented:
@rwjblue Interesting! Looks like that is related then, because if I re-enable `QUnit.config.autostart = false` and `QUnit.start()` from this addon, and comment them out in `ember-cli-qunit`, then tests run fine again. Having them both is problematic.

brzpegasus commented:
@sedouard Thank you very much!

sedouard commented:
:+1: :-)
